File: Sliding Window/424_LongestRepeatingCharacterReplacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        l = 0
        res = 0
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)
            if (r - l + 1) - max(count.values()) > k:
                count[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res


class Solution2:
    def characterReplacement(self, s: str, k: int) -> int:
        count = defaultdict(int)
        max_freq = 0
        l = 0
        res = 0
        for r in range(len(s)):
            count[s[r]] += 1
            max_freq = max(max_freq, count[s[r]])
            if (r - l + 1) - max_freq > k:
                count[s[l]] -= 1
                l += 1
            res = r - l + 1
        return res

File: Java Solutions/424_LongestRepeatingCharacterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] count = new int[26];
        int l = 0, res = 0;
        int maxFreq = 0;

        for (int r = 0; r < s.length(); r++) {
            int rChar = s.charAt(r) - 'A';
            count[rChar] += 1;
            maxFreq = Math.max(maxFreq, count[rChar]);
            if (r - l + 1 - maxFreq > k) {
                int lChar = s.charAt(l) - 'A';
                count[lChar] -= 1;
                l += 1;
            }
            res = r - l + 1;
        }

        return res;
    }
}
